### DISCUSSION TITLE
How to authenticate users with either TOTP OR Webauthn?

### DISCUSSION THREAD

**Discussion by lzc256** [keycloak/keycloak#21526] [Q&A] [answered]
Attempting to implement 2fa with Totp or webauthn, which means that: after the user's entering username and password, they can select whether to use TOTP or Webauthn as their 2fa authentication method, and the user would have access to the account if either TOTP or Webauthn authentication passed.

Having no idea how implement that. Any answer would be appreciated.

**Comment by christian-2** [comment, marked as answer]
The section on [W3C Web Authentication (WebAuthn)](https://www.keycloak.org/docs/latest/server_admin/index.html#webauthn_server_administration_guide) in the Keycloak [Server Administration Guide](https://www.keycloak.org/docs/latest/server_admin/index.html) and this [tutorial](https://keycloak.ch/keycloak-tutorials/tutorial-webauthn) are good starting points for the WebAuthn part. You'll find that this is not exactly a beginner's task, but that it will work well in the end.

**Reply by lzc256** [reply]
Thanks for replying!
